File: 7ReverseInteger.cpp
#include <iostream>
#include <set>
#include <vector>
#include <iterator>
#include <utility>
#include <cstdlib>
#include <limits>

using namespace std;

template<class Container>
void printContainer(const Container& container) {
    return;

    for (auto it = container.begin(); it != container.end(); it++) {
        if (it == container.begin()) {
            cout << "[";
        }
        cout << *it;
        if (it == prev(container.end())) {
            cout << "]" << endl;
        } else {
            cout << ", ";
        }
    }
}

class Solution {
public:
    int reverse(int x) {
        int sign = x < 0 ? -1 : 1;
        x = abs(x);
        uint32_t output = 0;

        while (x > 0) {
            int digit = x % 10;
            x = x / 10;

            uint32_t preOutput = output;
            output = output * 10 + digit;

            if ((output - digit) / 10 != preOutput ||
                (sign > 0 && output > static_cast<uint32_t>(numeric_limits<int>::max())) ||
                (sign < 0 && output > (static_cast<uint32_t>(numeric_limits<int>::max()) + 1))) {
                return 0;
            }
        }

        return sign * static_cast<int>(output);
    }
};

int main(){
    Solution s;
    auto output = s.reverse(123);
    auto output1 = s.reverse(-123);
    auto output2 = s.reverse(120);
    auto output3 = s.reverse(0);
    auto output4 = s.reverse(1534236469);

    cout << "*************output*************" << endl;
    cout << output << endl;
    cout << output1 << endl;
    cout << output2 << endl;
    cout << output3 << endl;
    cout << output4 << endl;

    return 0;
}

File: go/7ReverseInteger.go
package main

import (
	"container/list"
	"fmt"
	"math"
)

func reverse(x int) int {
	if x == 0 {
		return 0
	}

	var abs int
	var sign int
	if x >= 0 {
		abs = x
		sign = 1
	} else {
		abs = x * -1
		sign = -1
	}

	lstByte := list.New()
	for ; abs != 0; abs = abs / 10 {
		lstByte.PushBack(abs % 10)
	}

	var output int
	for e := lstByte.Front(); e != nil; e = e.Next() {
		output = output*10 + e.Value.(int)
		if (sign < 0 && output > math.MaxInt32+1) ||
			(sign > 0 && output > math.MaxInt32) {
			return 0
		}
	}

	return output * sign
}

func main() {
	output := reverse(123)
	output1 := reverse(-123)
	output2 := reverse(120)
	output3 := reverse(0)
	output4 := reverse(1534236469)

	fmt.Println("\n******** output *********\n", output, output1, output2, output3, output4)
}
